### PR TITLE
Save each evaluation epoch

### LINKED ISSUE #597: [FEATURE]when saving multiple epochs add an epoch number suffix for when save best=False
### 🚀 Feature

Saves multiple .pth on each checkpoint. Instead of overwriting every checkpoint.pth

### Motivation

Often useful to see how model performs at each epoch/savepoint. For example when training llm, want to measure the generative capabilities after each epoch and see if it is improving

### PR DESCRIPTION
**Pull request by pascal-pfeiffer**
New setting for "Save checkpoint"

When set to **Each evaluation epoch** it will save the model weights for each evaluation epoch. 
- This can be useful for debugging and experimenting, but will consume more disk space.
- Models uploaded to Hugging Face Hub will only contain the last checkpoint.
- Local downloads will contain all checkpoints.

closes #597

### CODE REVIEW

**Review by psinger — CHANGES_REQUESTED** [review]
Got the following error:

```
2024-05-24 11:12:16,173 - ERROR: Exception occurred during H2O LLM Studio run:
Traceback (most recent call last):
  File "/home/philipp/h2o-llmstudio/train_wave.py", line 106, in <module>
    run(cfg=cfg)
  File "/home/philipp/h2o-llmstudio/train.py", line 647, in run
    val_loss, val_metric = run_train(
  File "/home/philipp/h2o-llmstudio/train.py", line 393, in run_train
    create_symlinks_in_parent_folder(checkpoint_path)
  File "/home/philipp/h2o-llmstudio/llm_studio/src/utils/utils.py", line 176, in create_symlinks_in_parent_folder
    raise FileNotFoundError(f"Directory {directory} does not exist.")
FileNotFoundError: Directory /home/philipp/h2o-llmstudio/output/user/impartial-antelope.1.1/epoch_0_step_1073 does not exist.
2024-05-24 11:12:16,173 - ERROR: Exception occurred during H2O LLM Studio run:
Traceback (most recent call last):
  File "/home/philipp/h2o-llmstudio/train_wave.py", line 106, in <module>
    run(cfg=cfg)
  File "/home/philipp/h2o-llmstudio/train.py", line 647, in run
    val_loss, val_metric = run_train(
  File "/home/philipp/h2o-llmstudio/train.py", line 393, in run_train
    create_symlinks_in_parent_folder(checkpoint_path)
  File "/home/philipp/h2o-llmstudio/llm_studio/src/utils/utils.py", line 176, in create_symlinks_in_parent_folder
    raise FileNotFoundError(f"Directory {directory} does not exist.")
FileNotFoundError: Directory /home/philipp/h2o-llmstudio/output/user/impartial-antelope.1.1/epoch_0_step_1073 does not exist.
2024-05-24 11:12:16,177 - INFO: Saving model checkpoint to /home/philipp/h2o-llmstudio/output/user/impartial-antelope.1.1/epoch_0_step_1073
```

File: train.py [hunk 1]
```diff
@@ -371,32 +376,35 @@ def run_train(
 
             # Validation loop
             if (itr + 1) % evaluation_step == 0:
-                if cfg.training.evaluation_epochs == 1:
```
[the remark this answers is not in the document]
Comment (pascal-pfeiffer) [reply train.py:374]: I think it only prevents these from happening

```
2024-05-24 15:26:00,173 - INFO: train loss: 2.38: 40%|###9 | 1288/3223 [01:19<01:58, 16.30it/s]
2024-05-24 15:26:10,064 - INFO: train loss: 2.47: 45%|####4 | 1449/3223 [01:29<01:48, 16.30it/s]
2024-05-24 15:26:19,984 - INFO: train loss: 2.64: 50%|####9 | 1610/3223 [01:39<01:39, 16.27it/s]
2024-05-24 15:26:20,047 - INFO: Saving last model checkpoint to h2o-llmstudio/output/user/zircon-scallop/
2024-05-24 15:26:20,387 - INFO: Starting validation inference
2024-05-24 15:26:20,387 - INFO: validation progress: 0%| | 0/33 [00:00
2024-05-24 15:26:29,424 - INFO: validation progress: 3%|3 | 1/33 [00:09<04:49, 9.04s/it]
2024-05-24 15:26:30,787 - INFO: train loss: 2.64: 50%|####9 | 1610/3223 [01:50<01:39, 16.27it/s]
2024-05-24 15:26:31,317 - INFO: validation progress: 6%|6 | 2/33 [00:10<02:29, 4.83s/it]
2024-05-24 15:26:40,545 - INFO: validation progress: 9%|9 | 3/33 [00:20<03:25, 6.84s/it]
```

but only when `evaluation_epochs == 1`. 

The progress_bar is closed after each epoch anyway.

File: tests/integration/test_sequence_to_sequence_modeling_oasst_cpu_cfg.yaml [hunk 2]
```diff
@@ -73,6 +73,7 @@ tokenizer:
     max_length_answer: 16
     max_length_prompt: 16
     padding_quantile: 1.0
+    tokenizer_kwargs: '{"use_fast": true, "add_prefix_space": false}'
```
[the remark this answers is not in the document]
Comment (pascal-pfeiffer) [reply tests/integration/test_sequence_to_sequence_modeling_oasst_cpu_cfg.yaml:76]: Following [PEP20](https://peps.python.org/pep-0020/). "Explicit is better than implicit."

[the remark this answers is not in the document]
Comment (psinger) [reply tests/integration/test_sequence_to_sequence_modeling_oasst_cpu_cfg.yaml:76]: issue is that after each change all of these need to be updated as well
but fine

**Review by psinger — APPROVED** [review]
lgtm! ty

**Pull request merged**